controllers/util/provider.go

```diff
@@ -97,16 +96,30 @@ func encodeAndSignOnboardingToken(privateKeyPath string, ticket services.Onboard
 	return fmt.Sprintf("%s.%s", encodedPayload, encodedSignature), nil
 }
 
-func readAndDecodePrivateKey(privateKeyPath string) (*rsa.PrivateKey, error) {
-	pemString, err := os.ReadFile(privateKeyPath)
+func LoadOnboardingValidationPrivateKey(ctx context.Context, cl client.Client, namespace string) (*rsa.PrivateKey, error) {
+	privateSecret := &corev1.Secret{}
+	privateSecret.Name = onboardingValidationPrivateKeySecretName
+	privateSecret.Namespace = namespace
+
+	err := cl.Get(ctx, client.ObjectKeyFromObject(privateSecret), privateSecret)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read private key: %v", err)
+		return nil, fmt.Errorf("failed to get private secret: %v", err)
+	}
+
+	privateSecretKey := privateSecret.Data["key"]
+	if len(privateSecretKey) == 0 {
+		return nil, fmt.Errorf("No data found in secret")
+	}
+
+	Block, err1 := pem.Decode(privateSecretKey)
+	if err1 != nil {
+		return nil, fmt.Errorf("Failed to decode private key: %v", err1)
```

Comment:
lower case `Block` and as per `pem.Decode` func signature the second return value isn't err but a slice of bytes which should be empty.
```suggestion
	block, rest := pem.Decode(privateSecretKey)
	if len(rest) > 0 {
		return nil, fmt.Errorf("PEM block not found in private key: %s", rest)
```

Reply:
pem.Decode  returns byte, changed as suggested by Leela

Reply:
I also mentioned to lowercase `Block` as it doesn't have any extra significance.

Reply:
ack